scipy/constants/_codata.py

```diff
@@ -1496,40 +1529,186 @@
+        "joule-electron volt relationship": 1 / e,
+        "joule-hertz relationship": 1 / h,
+        "joule-inverse meter relationship": 1 / (h * c),
+        "joule-kelvin relationship": 1 / k,
+        "joule-kilogram relationship": 1 / c**2,
+        "kelvin-electron volt relationship": k / e,
+        "kelvin-hertz relationship": k / h,
+        "kelvin-inverse meter relationship": k / (h * c),
+        "kelvin-kilogram relationship": k / c**2,
+        "kilogram-electron volt relationship": c**2 / e,
+        "kilogram-hertz relationship": c**2 / h,
+        "kilogram-inverse meter relationship": c / h,
+        "kilogram-joule relationship": c**2,
+        "kilogram-kelvin relationship": c**2 / k,
+        "Loschmidt constant (273.15 K, 100 kPa)": 100e3 / 273.15 / k,
+        "Loschmidt constant (273.15 K, 101.325 kPa)": 101.325e3 / 273.15 / k,
+        "mag. flux quantum": h / (2 * e),
+        "molar gas constant": N_A * k,
+        "molar Planck constant": h * N_A,
+        'molar volume of ideal gas (273.15 K, 100 kPa)': N_A * k * 273.15 / 100e3,  # noqa: E501
+        'molar volume of ideal gas (273.15 K, 101.325 kPa)': N_A * k * 273.15 / 101.325e3,  # noqa: E501
+        "natural unit of action": (h / (2 * pi)),
+        "natural unit of action in eV s": (h / (2 * pi)) / e,
+        "Planck constant in eV/Hz": h / e,
+        "reduced Planck constant": h / (2 * pi),
+        "reduced Planck constant in eV s": (h / (2 * pi)) / e,
+        "reduced Planck constant times c in MeV fm": (h / (2*pi)) * c / (e * 1e6 * 1e-15),  # noqa: E501
+        "second radiation constant": h * c / k,
+        "Stefan-Boltzmann constant": 2 * pi**5 * k**4 / (15 * h**3 * c**2),
+        "von Klitzing constant": R_K,
+        "Wien frequency displacement law constant": alpha_W * k / h,
+        "Wien wavelength displacement law constant": h * c / (x_W * k),
+    }
+    return replace
 
 
-def parse_constants_2002to2014(d: str) -> dict[str, tuple[float, str, float]]:
-    constants = {}
+# -----------------------------------------------------------------------------
+
+
+def parse_constants_2002to2014(
+    d: str, exact_func: Callable[[Any], Any]
+) -> dict[str, tuple[float, str, float]]:
+    constants: dict[str, tuple[float, str, float]] = {}
+    exact: dict[str, float] = {}
+    need_replace = set()
     for line in d.split('\n'):
         name = line[:55].rstrip()
         val = float(line[55:77].replace(' ', '').replace('...', ''))
+        is_truncated = '...' in line[55:77]
+        is_exact = '(exact)' in line[77:99]
+        try:
+            if is_truncated and is_exact:
+                # missing decimals, use computed exact value
+                need_replace.add(name)
+            elif is_exact:
+                exact[name] = val
+            elif is_truncated:
+                raise Exception("Parsing error")
+        except Warning as w:
+            raise Warning(f"{name} is truncated but not exact.") from w
```

Comment:
About the exceptions, we should try to raise informative types of exceptions for the users, instead of `Warning`, it's more something like a `ValueError` that we might need.

For internal exceptions, we should also be more specific and not use a bare `Exception`. This to help prevent catching something else.

Lastly and this is a bit more work, we should test any new exception which we add. Our current policy is to do the following in Pytest:

```
message = r"Message in the exception"
with pytest.raises(ValueError, match=message):
    ...  # code that will raise
```

Reply:
Okay, I'll try to get to it at some other day.

Reply:
Thank you 🙇 No rush at all.